packages/engine/paima-funnel/src/funnels/carp/funnel.ts

```diff
@@ -0,0 +1,301 @@
+      const entry = sharedData.cacheManager.cacheEntries[CarpFunnelCacheEntry.SYMBOL];
+      if (entry != null) return entry;
+
+      const newEntry = new CarpFunnelCacheEntry();
+      sharedData.cacheManager.cacheEntries[CarpFunnelCacheEntry.SYMBOL] = newEntry;
+
+      newEntry.updateStartingSlot(
+        timestampToAbsoluteSlot(
+          (await sharedData.web3.eth.getBlock(startingBlockHeight)).timestamp as number
+        )
+      );
+
+      return newEntry;
+    })();
+
+    return new CarpFunnel(sharedData, dbTx, baseFunnel, carpUrl, await cacheEntry);
+  }
+}
+
+async function readDataInternal(
+  data: ChainData[],
+  carpUrl: string,
+  extensions: ChainDataExtension[],
+  lastTimestamp: number,
+  cache: CarpFunnelCacheEntry
+): Promise<PresyncChainData[]> {
+  // the lower range is exclusive
+  const min = timestampToAbsoluteSlot(lastTimestamp);
+  // the upper range is inclusive
+  const maxElement = data[data.length - 1];
+
+  const max = timestampToAbsoluteSlot(maxElement.timestamp);
+
+  cache.updateLastPoint(maxElement.blockNumber, maxElement.timestamp);
+
+  // Block finality depends on depth, and not on time, so it's possible that a
+  // block at a non confirmed depth falls in the slot range that we are querying
+  // here. This waits until the upper end of the range falls in the confirmed
+  // zone.
+  while (true) {
+    const stableBlock = await timeout(
+      query(carpUrl, Routes.blockLatest, {
+        offset: Number(confirmationDepth),
+      }),
+      DEFAULT_FUNNEL_TIMEOUT
+    );
+
+    if (stableBlock.block.slot > max) {
+      break;
+    }
+
+    await delay(delayForWaitingForFinalityLoop);
+  }
+
+  const blockNumbers = data.reduce(
+    (dict, data) => {
+      dict[timestampToAbsoluteSlot(data.timestamp)] = data.blockNumber;
+      return dict;
+    },
+    {} as { [slot: number]: number }
+  );
+
+  // This extends blockNumbers but for intermediate slots.
+  // Between two evm blocks there can be more than one slot, and the mapping only has the slots for blocks that exist.
+  const mapSlotToBlockNumber = (slot: number): number => {
+    while (true) {
+      const curr = blockNumbers[slot];
+      if (curr) {
+        return curr;
+      }
+      slot += 1;
+    }
+  };
+
+  const poolEvents = await Promise.all(
+    extensions.map((extension: ChainDataExtension): Promise<ChainDataExtensionDatum[]> => {
+      if ('stopSlot' in extension && extension.stopSlot && min >= extension.stopSlot) {
+        return Promise.resolve([]);
+      }
+
+      switch (extension.cdeType) {
+        case ChainDataExtensionType.CardanoPool:
+          const data = getCdePoolData(
+            carpUrl,
+            extension,
+            min,
+            Math.min(max, extension.stopSlot || max),
+            mapSlotToBlockNumber
+          );
+
+          return data;
+        default:
+          return Promise.resolve([]);
+      }
+    })
+  );
+
+  let grouped = groupCdeData(
+    Network.EVM,
```

Comment:
why not cardano?

Reply:
well... It doesn't actually matter, I think. 

The discriminant was introduced so that during the presync we can update the cardano cde tracking table instead of the other one, but during the normal sync just knowing the EVM block gives you the slot, so the field is not used.

But I guess we should put cardano here anyway to be consistent.